rllib/core/rl_module/marl_module.py

```diff
@@ -0,0 +1,295 @@
+            ValueError: If the module ID already exists and override is False.
+        """
+        if module_id in self._rl_modules and not override:
+            raise ValueError(
+                f"Module ID {module_id} already exists. If your intention is to "
+                "override, set override=True."
+            )
+        if self.is_distributed():
+            module.make_distributed(dist_config)
+        self._rl_modules[module_id] = module
+
+    def remove_module(
+        self, module_id: ModuleID, *, raise_err_if_not_found: bool = True
+    ) -> None:
+        """Removes a module at run time from the multi-agent module.
+
+        Args:
+            module_id: The module ID to remove.
+            raise_err_if_not_found: Whether to raise an error if the module ID is not
+                found.
+        Raises:
+            ValueError: If the module ID does not exist and raise_err_if_not_found is
+                True.
+        """
+        if raise_err_if_not_found:
+            self._check_module_exists(module_id)
+        del self._rl_modules[module_id]
+
+    def __getitem__(self, module_id: ModuleID) -> RLModule:
+        """Returns the module with the given module ID.
+
+        Args:
+            module_id: The module ID to get.
+
+        Returns:
+            The module with the given module ID.
+        """
+        self._check_module_exists(module_id)
+        return self._rl_modules[module_id]
+
+    @OverrideToImplementCustomLogic
+    def _make_modules(self) -> Mapping[ModuleID, RLModule]:
+        modules_info_dict = self.config["modules"]
+        modules = {}
+        for module_id, module_info in modules_info_dict.items():
+
+            if isinstance(module_info, RLModule):
+                modules[module_id] = module_info
+            elif isinstance(module_info, dict):
+                mod_class = module_info.get("module_class")
+                if mod_class is None:
+                    raise KeyError(
+                        f"key `module_class` is missing in the module "
+                        f"specfication of module {module_id}"
+                    )
+                mod_config = module_info.get("module_config", {})
+                modules[module_id] = mod_class(config=mod_config)
+            elif isinstance(module_id, tuple):
+                if len(module_id) != 2:
+                    raise ValueError(
+                        f"module {module_id} should has 2 elements: "
+                        f"(module_class, module_config)"
+                    )
+                mod_class, mod_config = module_info
+                modules[module_id] = mod_class(config=mod_config)
+            else:
+                raise ValueError(
+                    f"Invalid module info for module {module_id}: {module_info}"
+                )
+
+        return modules
+
+    @override(RLModule)
+    def output_specs_train(self) -> ModelSpec:
+        return self._get_specs_for_modules("output_specs_train")
```

Comment:
Dumb question: Where are these private methods defined?

Reply:
which one? `_get_specs_for_modules`? at the bottom of these spec defs? "output_specs_train" for example is also defined in the underlying RLModules, so you are just looping through RLModules and getting their expected format and map them to their ModuleIDs?